common/flatpak-dir.c

```diff
@@ -15269,7 +15258,34 @@ flatpak_dir_find_local_related_for_metadata (FlatpakDir        *self,
                 }
               else if (subdirectories)
                 {
-                  g_autoptr(GHashTable) matches = local_match_prefix (self, extension_ref, remote_name);
+                  g_autoptr(GHashTable) matches = NULL;
+
+                  if (!all_refs_for_remote)
+                    {
+                      g_autoptr(GHashTable) full_refs = NULL;
+                      g_autofree char *list_prefix = NULL;
+                      if (remote_name != NULL)
+                        list_prefix = g_strdup_printf ("%s:", remote_name);
+
+                      if (ostree_repo_list_refs (self->repo, list_prefix, &full_refs, NULL, NULL))
+                        {
+                          GHashTableIter iter;
+                          gpointer key;
+
+                          all_refs_for_remote = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+                          g_hash_table_iter_init(&iter, full_refs);
+                          while (g_hash_table_iter_next (&iter, &key, NULL))
+                            {
+                              const char *partial_ref_and_origin = key;
+                              g_autofree char *partial_ref = NULL;
+                              if (ostree_parse_refspec (partial_ref_and_origin, NULL, &partial_ref, NULL))
+                                g_hash_table_add (all_refs_for_remote, g_steal_pointer (&partial_ref));
```

Comment:
While this seems correct, it isn't really the best way to do this. I think you should run flatpak_decomposed_new_from_ref() here, and store the FlatpakDecomposed in the hash table. This allows for several things: 
 * No need to decompose everything multiple times
 * Don't store invalid flatpak refs in the hash (for example, the deploy refs if we're not using a remote prefix to pre-filter.
 * Simpler matching on the kind in local_match_prefix

Reply:
Just pushed a new version following that style, it's definitely *much* nicer.